_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+```sh
+cd ~/git
+mkdir rails
+cd rails
+git clone git@github.com:rails/rails.git
+cd rails
+```
+
+I searched for `set_autoload_path` using `ack`, which is a nicer alternative to `grep`:
+
+```sh
+git checkout v3.2.2
+ack set_autoload_paths
+```
+
+I saw only 2 results
+
+![ack set_autoload_paths](/images/ack-set_autoload_paths.png)
+
+`railties/lib/rails/engine.rb` line 536 looks like a DSL call for setting up an initializer, so I opened `~/git/rails/rails/railties` in Rubymine to take advantage of its ability to jump from definition and usage easier than vim.  I did Shift+Shift and typed `engine.rb` and hit enter to open the file.  I did `:536` to jump to line `536` because I use IDEAVim to have vim bindings in Rubymine.
+
+![railties/lib/rails/engine.rb:536](/images/railties-lib-rails-engine-line-536.png)
+
+### `set_autoload_paths` implementation
+
+I see `ActiveSupport::Dependencies.autoload_paths` being added to with `unshift`, but I don't see `config.autoload_paths` being read except for the `config.autoload_paths.freeze` on line 541, so I Cmd+Click `_all_autoload_paths` to see it's definition.
+
+![all_autoload_paths](/images/all_autoload_paths.png)
+
+`_all_autoload_paths` is just the combination of `config.autoload_paths`, `config.eager_load_paths`, and `config.autoload_once_paths`, so if all I care about is emulating `set_autoload_paths` is I need to do `ActiveSupport::Dependencies.autoload_paths.unshift(config.autoload_paths)`, which is what I ended up doing in the [PR](https://github.com/rapid7/pro/blob/841f0d7c38d9fbb99ea8df742d15f07d32e55e3f/engine/lib/metasploit/configured.rb#L58).  Looking at the other lines in `set_autoload_paths`, I'll want to [`freeze` my paths to prevent future modifications](https://github.com/rapid7/pro/blob/841f0d7c38d9fbb99ea8df742d15f07d32e55e3f/engine/lib/metasploit/configured.rb#L60-L61).
+
+### Emulating initializers without `Rails::Engine`
+
+So, that's take care of what my `set_autoload_paths` should do, but how am I going to get prosvc to do it without `Rails::Engine`?  I need to look more at how `Rails::Engine`s works to see which pieces I should copy to mimic its path handling from boot to set_autoload_paths.  I know that initializer are run in the order they are defined except where `:before` or `:after` would override that order from [the configuring guide](http://guides.rubyonrails.org/v3.2.2/configuring.html#rails-railtie-initializer).
```

Comment:
So, that's take care of what my `set_autoload_paths` should do, but how am I going to get prosvc to do it without `Rails::Engine`?  I need to look more at how `Rails::Engine`s works to see which pieces I should copy to mimic its path handling from boot to set_autoload_paths.  I know that **~~initializer~~** **initializers** are run in the order they are defined except where `:before` or `:after` would override that order from [the configuring guide](http://guides.rubyonrails.org/v3.2.2/configuring.html#rails-railtie-initializer).